In [425]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

##### Load the csv files

In [426]:
car_data = pd.read_csv('car data.csv')
cardekho = pd.read_csv('CAR DETAILS FROM CAR DEKHO.csv')
car_detail = pd.read_csv('Car details v3.csv')


In [427]:
car_data.head()

Car_Name  Year  Selling_Price  Present_Price  Kms_Driven Fuel_Type  \
0     ritz  2014           3.35           5.59       27000    Petrol   
1      sx4  2013           4.75           9.54       43000    Diesel   
2     ciaz  2017           7.25           9.85        6900    Petrol   
3  wagon r  2011           2.85           4.15        5200    Petrol   
4    swift  2014           4.60           6.87       42450    Diesel   

  Seller_Type Transmission  Owner  
0      Dealer       Manual      0  
1      Dealer       Manual      0  
2      Dealer       Manual      0  
3      Dealer       Manual      0  
4      Dealer       Manual      0

In [428]:
car_data.shape

(301, 9)

In [429]:
cardekho.head()

name  year  selling_price  km_driven    fuel  \
0             Maruti 800 AC  2007          60000      70000  Petrol   
1  Maruti Wagon R LXI Minor  2007         135000      50000  Petrol   
2      Hyundai Verna 1.6 SX  2012         600000     100000  Diesel   
3    Datsun RediGO T Option  2017         250000      46000  Petrol   
4     Honda Amaze VX i-DTEC  2014         450000     141000  Diesel   

  seller_type transmission         owner  
0  Individual       Manual   First Owner  
1  Individual       Manual   First Owner  
2  Individual       Manual   First Owner  
3  Individual       Manual   First Owner  
4  Individual       Manual  Second Owner

In [430]:
cardekho.shape

(4340, 8)

In [431]:
car_detail.head(10)

name  year  selling_price  km_driven  \
0                Maruti Swift Dzire VDI  2014         450000     145500   
1          Skoda Rapid 1.5 TDI Ambition  2014         370000     120000   
2              Honda City 2017-2020 EXi  2006         158000     140000   
3             Hyundai i20 Sportz Diesel  2010         225000     127000   
4                Maruti Swift VXI BSIII  2007         130000     120000   
5         Hyundai Xcent 1.2 VTVT E Plus  2017         440000      45000   
6          Maruti Wagon R LXI DUO BSIII  2007          96000     175000   
7                    Maruti 800 DX BSII  2001          45000       5000   
8                      Toyota Etios VXD  2011         350000      90000   
9  Ford Figo Diesel Celebration Edition  2013         200000     169000   

     fuel seller_type transmission         owner     mileage   engine  \
0  Diesel  Individual       Manual   First Owner   23.4 kmpl  1248 CC   
1  Diesel  Individual       Manual  Second Owner  21.14 kmpl  1498 CC   
2  Petrol  Individual       Manual   Third Owner   17.7 kmpl  1497 CC   
3  Diesel  Individual       Manual   First Owner   23.0 kmpl  1396 CC   
4  Petrol  Individual       Manual   First Owner   16.1 kmpl  1298 CC   
5  Petrol  Individual       Manual   First Owner  20.14 kmpl  1197 CC   
6     LPG  Individual       Manual   First Owner  17.3 km/kg  1061 CC   
7  Petrol  Individual       Manual  Second Owner   16.1 kmpl   796 CC   
8  Diesel  Individual       Manual   First Owner  23.59 kmpl  1364 CC   
9  Diesel  Individual       Manual   First Owner   20.0 kmpl  1399 CC   

    max_power                    torque  seats  
0      74 bhp            190Nm@ 2000rpm    5.0  
1  103.52 bhp       250Nm@ 1500-2500rpm    5.0  
2      78 bhp     12.7@ 2,700(kgm@ rpm)    5.0  
3      90 bhp  22.4 kgm at 1750-2750rpm    5.0  
4    88.2 bhp     11.5@ 4,500(kgm@ rpm)    5.0  
5   81.86 bhp         113.75nm@ 4000rpm    5.0  
6    57.5 bhp      7.8@ 4,500(kgm@ rpm)    5.0  
7      37 bhp             59Nm@ 2500rpm    4.0  
8    67.1 bhp       170Nm@ 1800-2400rpm    5.0  
9    68.1 bhp            160Nm@ 2000rpm    5.0

In [432]:
car_detail.shape

(8128, 13)

All the files can be used for Rregression but as the last CSV has highest number of Datapoints I will be using that to train my model

##### EDA

In [433]:
car_detail.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8128 entries, 0 to 8127
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           8128 non-null   object 
 1   year           8128 non-null   int64  
 2   selling_price  8128 non-null   int64  
 3   km_driven      8128 non-null   int64  
 4   fuel           8128 non-null   object 
 5   seller_type    8128 non-null   object 
 6   transmission   8128 non-null   object 
 7   owner          8128 non-null   object 
 8   mileage        7907 non-null   object 
 9   engine         7907 non-null   object 
 10  max_power      7913 non-null   object 
 11  torque         7906 non-null   object 
 12  seats          7907 non-null   float64
dtypes: float64(1), int64(3), object(9)
memory usage: 825.6+ KB


So there are null values present, these I will handle in Feature engineering step

In [434]:
car_detail.isnull().sum()

name               0
year               0
selling_price      0
km_driven          0
fuel               0
seller_type        0
transmission       0
owner              0
mileage          221
engine           221
max_power        215
torque           222
seats            221
dtype: int64

In [435]:
#let's check the model of the cars
print(car_detail['name'].unique()[1:100])
print(len(car_detail['name'].unique()))
#(Word of caution) using OHE will create aditional 2058 columns

['Skoda Rapid 1.5 TDI Ambition' 'Honda City 2017-2020 EXi'
 'Hyundai i20 Sportz Diesel' 'Maruti Swift VXI BSIII'
 'Hyundai Xcent 1.2 VTVT E Plus' 'Maruti Wagon R LXI DUO BSIII'
 'Maruti 800 DX BSII' 'Toyota Etios VXD'
 'Ford Figo Diesel Celebration Edition' 'Renault Duster 110PS Diesel RxL'
 'Maruti Zen LX' 'Maruti Swift Dzire VDi' 'Maruti Swift 1.3 VXi'
 'Maruti Wagon R LXI Minor' 'Mahindra KUV 100 mFALCON G80 K8 5str'
 'Maruti Ertiga SHVS VDI' 'Hyundai i20 1.4 CRDi Asta' 'Maruti Alto LX'
 'Hyundai i20 2015-2017 Asta 1.4 CRDi' 'Mahindra Verito 1.5 D4 BSIII'
 'Honda WR-V i-DTEC VX' 'Maruti Swift Dzire ZDI' 'Maruti SX4 ZDI'
 'Tata Tigor 2017-2020 XZ' 'Maruti Baleno Delta 1.2' 'Maruti Alto 800 LXI'
 'Chevrolet Enjoy TCDi LTZ 7 Seater' 'Maruti Omni E MPI STD BS IV'
 'Maruti Vitara Brezza LDi Option' 'Fiat Palio 1.2 ELX'
 'Maruti Omni 8 Seater BSIV' 'Hyundai i20 1.2 Asta Dual Tone'
 'Maruti Vitara Brezza LDi' 'Maruti Alto 800 CNG LXI Optional'
 'Maruti Alto K10 VXI Airbag' 'Hyundai Verna V

In [436]:
print(car_detail['year'].unique())
print(len(car_detail['year'].unique()))
#does not need any encoding

[2014 2006 2010 2007 2017 2001 2011 2013 2005 2009 2016 2012 2002 2015
 2018 2003 2019 2008 2020 1999 2000 1983 2004 1996 1994 1995 1998 1997
 1991]
29


In [437]:
print(car_detail['torque'].unique())
print(len(car_detail['torque'].unique()))


['190Nm@ 2000rpm' '250Nm@ 1500-2500rpm' '12.7@ 2,700(kgm@ rpm)'
 '22.4 kgm at 1750-2750rpm' '11.5@ 4,500(kgm@ rpm)' '113.75nm@ 4000rpm'
 '7.8@ 4,500(kgm@ rpm)' '59Nm@ 2500rpm' '170Nm@ 1800-2400rpm'
 '160Nm@ 2000rpm' '248Nm@ 2250rpm' '78Nm@ 4500rpm' nan '84Nm@ 3500rpm'
 '115Nm@ 3500-3600rpm' '200Nm@ 1750rpm' '62Nm@ 3000rpm'
 '219.7Nm@ 1500-2750rpm' '114Nm@ 3500rpm' '115Nm@ 4000rpm' '69Nm@ 3500rpm'
 '172.5Nm@ 1750rpm' '6.1kgm@ 3000rpm' '114.7Nm@ 4000rpm' '60Nm@ 3500rpm'
 '90Nm@ 3500rpm' '151Nm@ 4850rpm' '104Nm@ 4000rpm' '320Nm@ 1700-2700rpm'
 '250Nm@ 1750-2500rpm' '145Nm@ 4600rpm' '146Nm@ 4800rpm'
 '343Nm@ 1400-3400rpm' '200Nm@ 1400-3400rpm' '200Nm@ 1250-4000rpm'
 '400Nm@ 2000-2500rpm' '138Nm@ 4400rpm' '360Nm@ 1200-3400rpm'
 '200Nm@ 1200-3600rpm' '380Nm@ 1750-2500rpm' '173Nm@ 4000rpm'
 '400Nm@ 1750-3000rpm' '400Nm@ 1400-2800rpm' '200Nm@ 1750-3000rpm'
 '111.7Nm@ 4000rpm' '219.6Nm@ 1500-2750rpm' '112Nm@ 4000rpm'
 '250Nm@ 1500-3000rpm' '130Nm@ 4000rpm' '205Nm@ 1750-3250rpm'
 '280Nm@ 1350-46

**This is a problemetic feature , there are different formats and units used, Also the horse power and engine size has some what linear relationship with torque, so it can be dropped**

In [438]:
print(car_detail['seats'].unique())
print(len(car_detail['seats'].unique()))
# missing values are in form of np.nan

[ 5.  4. nan  7.  8.  6.  9. 10. 14.  2.]
10


In [439]:
print(car_detail['owner'].unique())
print(len(car_detail['owner'].unique()))
# missing values are in form of np.nan

['First Owner' 'Second Owner' 'Third Owner' 'Fourth & Above Owner'
 'Test Drive Car']
5


**Results of EDA:**
- Feature_mv__Action
- name              0 : OHE
- year              0 : None
- selling_price     0 : Target
- km_driven         0 : None
- fuel              0 : OHE
- seller_type       0 : OHE
- transmission      0 : OHE
- owner             0 : ORDINAL ENCODING
- mileage          221: Impute
- engine           221: Impute
- max_power        215: Impute
- torque           222: Drop
- seats            221: Impute

##### Feature Engineeering

First I will handle the units associated with columns

In [440]:
car_detail['mileage'] = car_detail['mileage'].str.replace('kmpl','')
car_detail['mileage'] = car_detail['mileage'].str.replace('km/kg','')
car_detail['mileage'] = car_detail['mileage'].astype('float64')

car_detail['engine'] = car_detail['engine'].str.replace('CC','')
car_detail['engine'] = car_detail['engine'].astype('float64')

#car_detail['max_power'] = car_detail['max_power'].str.replace(' bhp','')
#car_detail['max_power'] = car_detail['max_power'].str.replace(' hp','')
#car_detail['max_power'] = car_detail['max_power'].str.replace(' PS','')
#car_detail['max_power'] = car_detail['max_power'].astype('float64')

car_detail['max_power'] = car_detail['max_power'].str.extract(r'(\d+.\d+)').astype('float')
car_detail['max_power'] = car_detail['max_power'].astype('float64')

In [441]:
car_detail.describe()

year  selling_price     km_driven      mileage       engine  \
count  8128.000000   8.128000e+03  8.128000e+03  7907.000000  7907.000000   
mean   2013.804011   6.382718e+05  6.981951e+04    19.418783  1458.625016   
std       4.044249   8.062534e+05  5.655055e+04     4.037145   503.916303   
min    1983.000000   2.999900e+04  1.000000e+00     0.000000   624.000000   
25%    2011.000000   2.549990e+05  3.500000e+04    16.780000  1197.000000   
50%    2015.000000   4.500000e+05  6.000000e+04    19.300000  1248.000000   
75%    2017.000000   6.750000e+05  9.800000e+04    22.320000  1582.000000   
max    2020.000000   1.000000e+07  2.360457e+06    42.000000  3604.000000   

         max_power        seats  
count  5994.000000  7907.000000  
mean     98.001464     5.416719  
std      38.308191     0.959588  
min      32.800000     2.000000  
25%      73.900000     5.000000  
50%      88.700000     5.000000  
75%     115.000000     5.000000  
max     400.000000    14.000000

In [442]:

X,y = car_detail,car_detail['selling_price']
X.drop(columns=['selling_price','torque'])

name  year  km_driven    fuel seller_type  \
0           Maruti Swift Dzire VDI  2014     145500  Diesel  Individual   
1     Skoda Rapid 1.5 TDI Ambition  2014     120000  Diesel  Individual   
2         Honda City 2017-2020 EXi  2006     140000  Petrol  Individual   
3        Hyundai i20 Sportz Diesel  2010     127000  Diesel  Individual   
4           Maruti Swift VXI BSIII  2007     120000  Petrol  Individual   
...                            ...   ...        ...     ...         ...   
8123             Hyundai i20 Magna  2013     110000  Petrol  Individual   
8124         Hyundai Verna CRDi SX  2007     119000  Diesel  Individual   
8125        Maruti Swift Dzire ZDi  2009     120000  Diesel  Individual   
8126               Tata Indigo CR4  2013      25000  Diesel  Individual   
8127               Tata Indigo CR4  2013      25000  Diesel  Individual   

     transmission                 owner  mileage  engine  max_power  seats  
0          Manual           First Owner    23.40  1248.0        NaN    5.0  
1          Manual          Second Owner    21.14  1498.0     103.52    5.0  
2          Manual           Third Owner    17.70  1497.0        NaN    5.0  
3          Manual           First Owner    23.00  1396.0        NaN    5.0  
4          Manual           First Owner    16.10  1298.0      88.20    5.0  
...           ...                   ...      ...     ...        ...    ...  
8123       Manual           First Owner    18.50  1197.0      82.85    5.0  
8124       Manual  Fourth & Above Owner    16.80  1493.0     110.00    5.0  
8125       Manual           First Owner    19.30  1248.0      73.90    5.0  
8126       Manual           First Owner    23.57  1396.0        NaN    5.0  
8127       Manual           First Owner    23.57  1396.0        NaN    5.0  

[8128 rows x 11 columns]

In [443]:
y.head()

0    450000
1    370000
2    158000
3    225000
4    130000
Name: selling_price, dtype: int64

In [444]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2)

Here I will dcreate the Column transformer object to handle the missing and catagorical values

In [445]:
X_train['owner'].unique()

array(['First Owner', 'Second Owner', 'Fourth & Above Owner',
       'Third Owner', 'Test Drive Car'], dtype=object)

In [446]:
from sklearn.compose import ColumnTransformer
from sklearn.impute import KNNImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import OrdinalEncoder
ct = ColumnTransformer(transformers= [
('si',KNNImputer(),['mileage','engine','max_power','seats']),
('oe',OrdinalEncoder(categories='auto'),['owner']),
('ohe',OneHotEncoder(sparse=False,drop='first',handle_unknown='ignore'),['name','seller_type','transmission','fuel'])
])

In [447]:
X_letc = ct.fit_transform(X_train)
X_letc.shape

(6502, 1894)

In [448]:
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
lr = LinearRegression()
pipe = Pipeline([('ct',ct),('scaler',StandardScaler()),('lr',lr)])
pipe.fit(X_train,y_train)
y_pred = pipe.predict(X_test);

c:\Users\dsudi\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\preprocessing\_encoders.py:188: UserWarning: Found unknown categories in columns [0] during transform. These unknown categories will be encoded as all zeros
  warnings.warn(


In [449]:
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error
print(r2_score(y_test,y_pred))
print(mean_absolute_error(y_test,y_pred))

-1.3382605319768656e+23
4.4810709696197784e+16


Linear regression gave very bad result,data might not be linearly distributed let's changing it with XGBregressor

In [450]:
from xgboost import XGBRegressor
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
xgb = XGBRegressor(booster = 'gblinear',n_estimators = 500)
pipe = Pipeline([('ct',ct),('scaler',StandardScaler()),('xg',xgb)])
pipe.fit(X_train,y_train)
y_pred = pipe.predict(X_test);

c:\Users\dsudi\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\preprocessing\_encoders.py:188: UserWarning: Found unknown categories in columns [0] during transform. These unknown categories will be encoded as all zeros
  warnings.warn(


In [451]:

from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error
print(r2_score(y_test,y_pred))
print(mean_absolute_error(y_test,y_pred))

0.9227028514742395
106231.08913324973


In [452]:
n,p = X_train.shape
def adjr2(actual, predicted, rowcount, featurecount):
    
    return 1-(1-r2_score(actual,predicted))*(rowcount-1)/(rowcount-featurecount)
adjr2(y_test,y_pred,n,p)

0.9225599071403962

Let's see if removing outliers helps

In [453]:
columns = X_train.columns
for c in columns:
    if X_train[c].dtype == 'object':
        continue
    q1 = X_train[c].quantile(0.25)
    q3 = X_train[c].quantile(0.75)
    iqr = q3 - q1
    min = q1 - 1.5 * iqr
    max = q3 + 1.5 * iqr
    X_train.loc[X_train[c] < min, c] = min
    X_train.loc[X_train[c] > max, c] = max

In [454]:
from xgboost import XGBRegressor
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
lr = XGBRegressor(booster = 'gblinear')
pipe = Pipeline([('ct',ct),('scaler',StandardScaler()),('lr',lr)])
pipe.fit(X_train,y_train)
y_pred = pipe.predict(X_test);

c:\Users\dsudi\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\preprocessing\_encoders.py:188: UserWarning: Found unknown categories in columns [0] during transform. These unknown categories will be encoded as all zeros
  warnings.warn(


In [455]:

from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error
print(r2_score(y_test,y_pred))
print(mean_absolute_error(y_test,y_pred))
print(adjr2(y_test,y_pred,n,p))

0.8936213581177352
128400.76719254497
0.8934246338609025


Did not help, So I will stick with the first model